geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryMonitor.java

```diff
@@ -229,95 +197,89 @@ public static long getMemoryUsedDuringLowMemory() {
+     * due to low memory.
+     */
+    cancellingDueToLowMemory = true;
 
-  private void cancelQueryDueToLowMemory(QueryThreadTask queryTask, long memoryThreshold) {
-    boolean[] queryCompleted = queryTask.query.getQueryCompletedForMonitoring();
-    synchronized (queryCompleted) {
-      if (!queryCompleted[0]) {
-        // cancel if query is not completed
-        String reason = String.format(
-            "Query execution canceled due to memory threshold crossed in system, memory used: %s bytes.",
-            memoryThreshold);
-        queryTask.query.setCanceled(
-            new QueryExecutionLowMemoryException(reason));
-        queryTask.queryCancelled.set(Boolean.TRUE);
+    try {
+      /*
+       * It's tempting to try to process the list of tasks returned from shutdownNow().
+       * Unfortunately, that call leaves the executor in a state that causes the task's
+       * run() to cancel the task, instead of actually running it. By calling shutdown()
+       * we block new task additions and put the executor in a state that allows the
+       * task's run() to actually run the task logic.
+       */
+      executor.shutdown(); // executor won't accept new work ever again
+      final BlockingQueue<Runnable> expirationTaskQueue = executor.getQueue();
+      for (final Runnable cancellationTask : expirationTaskQueue) {
+        cancellationTask.run();
       }
+    } finally {
+      cancellingDueToLowMemory = false;
     }
-  }
 
-  /** FOR TEST PURPOSE */
-  public static int getQueryMonitorThreadCount() {
-    return queryThreads.size();
   }
 
-  /**
-   * Query Monitoring task, placed in the queue.
-   */
-  private static class QueryThreadTask {
-
-    // package-private to avoid synthetic accessor
-    final long StartTime;
+  private ScheduledFuture<?> scheduleCancellationTask(final DefaultQuery query,
+      final long timeLimitMillis) {
 
-    // package-private to avoid synthetic accessor
-    final Thread queryThread;
+    // make ThreadLocal QueryCancelled, available to closure, which will run in a separate thread
+    final AtomicBoolean queryCancelledThreadLocal =
+        DefaultQuery.QueryCancelled.get();
 
-    // package-private to avoid synthetic accessor
-    final DefaultQuery query;
+    return executor.schedule(() -> {
+      final CacheRuntimeException exception = cancellingDueToLowMemory ? createLowMemoryException()
+          : createExpirationException(timeLimitMillis);
 
-    // package-private to avoid synthetic accessor
-    final AtomicBoolean queryCancelled;
+      query.setQueryCancelledException(exception);
+      queryCancelledThreadLocal.set(true);
 
-    QueryThreadTask(Thread queryThread, DefaultQuery query, AtomicBoolean queryCancelled) {
-      this.StartTime = System.currentTimeMillis();
-      this.queryThread = queryThread;
-      this.query = query;
-      this.queryCancelled = queryCancelled;
-    }
+      logger.info(exception.getMessage());
```

Comment:
We don't really have any identifying information in our exception message regarding which query is actually being cancelled.  In the original code run() loop, we logged the QueryThreadTask.toString(), which included the queryThread and queryCancelled.  I'm not sure if that message was providing any value, but I presume if we log anything at all here it should be able to tell us _which_ query is being cancelled.